42 Improve Token Security & Extend `wac.install`

type1fool opened the pull request:
# Overview

Resolves #42
Resolves #41 

This PR contains two major changes:

- Tokens are stored in session cookies instead of `window.sessionStorage`.
- The `wac.install` Mix Task has been extended to generate more files and inject routes into the app's router.

## Changes

- Replaced `TokenComponent` with `Session` controller
- Updated `wac.install` to add options for bypassing generators
- Updated generators:
  - Contexts
  - Fixtures
  - Migrations
  - Schemas
  - Tests
- Added generators:
  - Controllers
  - LiveViews
  - Router
  - Session Hooks
- Moved schemas into `Identity` context
- Updated JS hooks to use ESM syntax
- Bumped to v0.6.0

## Video Demo

https://youtu.be/GGWM_ZM3xlg

## Tests

**WebauthnComponents**

```elixir
mix test

...EVENTS_OMITTED...

Finished in 0.1 seconds (0.1s async, 0.00s sync)
10 tests, 0 failures

Randomized with seed 373636
```

**Fresh Demo App w/ WAC**

```elixir
mix test
.......................
Finished in 0.2 seconds (0.2s async, 0.00s sync)
23 tests, 0 failures

Randomized with seed 313260
```

## UI Screenshots

The following screenshots were captured in Chrome on MacOS.

<details>
  <summary><h3>/sign-in</h3></summary>

<img width="422" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/7ae80982-8ece-47c3-a8a8-6b022905eb5a">

<img width="416" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/dd84da02-e30e-46d1-afdf-5b9dfc0e9d52">

</details>


<details>
  <summary><h3>Validation</h3></summary>

**Invalid Email**

<img width="423" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/a268dfbb-41c1-4068-ab3f-64d99d7286e6">

**Valid Email**

<img width="417" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/55595d0a-44ee-48f5-bbfe-c849425c13c6">

</details>

<details>
  <summary><h3>Registration</h3></summary>

<img width="496" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/33649d3d-fa0b-487b-af51-887b2df23eea">

<img width="262" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/68cb5c12-f81b-42a1-b82a-2d6cac7c3e05">

<img width="497" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/88a5b132-7cf4-455c-a14b-22382e1b1290">

</details>

<details>
<summary><h3>Authentication</h3></summary>

<img width="496" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/1d71113c-599e-4bda-a17a-14383d72bdbf">

<img width="498" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/17f0507c-3897-4dfb-b03f-5f1cdf04de06">

<img width="263" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/f75e14f9-a841-4da9-8558-8d9e46efbd12">

<img width="500" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/148e7237-6009-4ce1-a86d-f3998670121d">

</details>

<details>
<summary><h3>🚨 Firefox</h3></summary>
Firefox does not appear to support MacOS Passkeys by default. After attempting to register a user in Firefox, the browser prompted to activate the authenticator, but nothing happened when I pressed the laptop's TouchID.

<img width="780" alt="image" src="https://github.com/liveshowy/webauthn_components/assets/13895134/53c240f5-0a43-4192-9376-3f85b6d5c4be">

</details>

## Collaborators

1. @type1fool
1. @doorgan

type1fool commented:
## Example Success Output

```
mix wac.install
==> webauthn_components
Compiling 1 file (.ex)
==> passkeys
* creating lib/passkeys/users.ex
* creating lib/passkeys/users/user.ex
* creating lib/passkeys/users/user_key.ex
* creating lib/passkeys/users/user_token.ex
* creating priv/repo/migrations/20230913013459_users.exs
* creating priv/repo/migrations/20230913013500_user_keys.exs
* creating priv/repo/migrations/20230913013501_user_tokens.exs
* creating test/passkeys/users_test.exs
* creating test/support/user_fixtures.ex
* creating lib/passkeys_web/controllers/session.ex
* creating lib/passkeys_web/session_hooks/assign_user.ex
* creating lib/passkeys_web/session_hooks/require_user.ex
* creating lib/passkeys_web/live/authentication_live.ex
* creating lib/passkeys_web/live/authentication_live.html.heex
* updating lib/passkeys_web/router.ex

✅ Successfully scaffolded WebauthnComponents for Elixir.Passkeys

📚 Resources

- Repo: https://github.com/liveshowy/webauthn_components
- Hex:  https://hex.pm/packages/webauthn_components
- Docs: https://hexdocs.pm/webauthn_components/readme.html
```

## Example Error Output

```
mix wac.install
==> webauthn_components
Compiling 1 file (.ex)
==> passkeys
* creating lib/passkeys/users.ex
* creating lib/passkeys/users/user.ex
* creating lib/passkeys/users/user_key.ex
* creating lib/passkeys/users/user_token.ex
* creating priv/repo/migrations/20230913012748_users.exs
* creating priv/repo/migrations/20230913012749_user_keys.exs
* creating priv/repo/migrations/20230913012750_user_tokens.exs
* creating test/passkeys/users_test.exs
* creating test/support/user_fixtures.ex
* creating lib/passkeys_web/controllers/session.ex
* creating lib/passkeys_web/session_hooks/assign_user.ex
* creating lib/passkeys_web/session_hooks/require_user.ex
* creating lib/passkeys_web/live/authentication_live.ex
* creating lib/passkeys_web/live/authentication_live.html.heex
** (Wac.Gen.RouterError) 🫣 Unable to find the first router scope
🙏 Please review the WebauthnComponents issue tracker and open a new issue if necessary.
👉 https://github.com/liveshowy/webauthn_components/issues
👇 For debugging, please include this error and the stacktrace below:

    (webauthn_components 0.6.0) lib/wac_gen/router.ex:94: Wac.Gen.Router.insert_routes/2
    (webauthn_components 0.6.0) lib/wac_gen/router.ex:51: Wac.Gen.Router.update_router/1
    (mix 1.15.4) lib/mix/task.ex:447: anonymous fn/3 in Mix.Task.run_task/5
    (mix 1.15.4) lib/mix/cli.ex:92: Mix.CLI.run_task/2
    /Users/owenbickford/.asdf/installs/elixir/1.15.4-otp-26/bin/mix:2: (file)
    (elixir 1.15.4) lib/code.ex:1435: Code.require_file/2
```

houllette commented:
Hey @type1fool, great work on this and I'm excited to see this component library continue to evolve!

I reviewed this PR (and by extension the library itself) through a Security lens to try and identify vulnerabilities / general insecurities since you had requested a review in the #security channel of the Elixir language Slack server.

Overall, it's looking great! I have a couple of small findings that are worth mentioning along with some things I was slightly confused about. Please note that I am coming into your codebase extremely fresh, so there's always a chance that my findings are non-issues because of something I may have overlooked / not considered 🙂 

## Things that _may_ be issues
### Session Tokens Stored in Plaintext
Currently once the tokens are generated using `:crypto.strong_rand_bytes(64)` they're being stored as the raw binary in database. This primarily becomes an issue if the database is improperly accessed (leak, internal abuse, etc.) since _technically_ these are valid session tokens and in some circumstances could be used to perform user actions (now in their current form they couldn't be used like that due to cookie signing and whatnot, but it still doesn't negate the fact that they are valid user tokens being stored in the db)

A fix to this issue would be to put the tokens through a one-way hashing function before storing them in the DB and then on token queries you would also need to hash the token before DB lookup.

> Fwiw I went back and forth on including this one. There's [this discussion](https://elixirforum.com/t/why-does-phx-gen-auth-not-hash-session-tokens/47769) with a response from José on why `phx_gen_auth` _doesn't_ hash session tokens, but then there's [this](https://security.stackexchange.com/questions/221841/hashing-session-id) more general InfoSec discussion on why it may be worthwhile.
>
> At the end of the day it isn't necessarily insecure leaving it as is because you're inheriting the security of cookie signing from `Plug.Session`; but I will say that if someone (for whatever reason) chose to overwrite that functionality with their own or used another unknown module that didn't do signing, then you'd lose your safety net. 
>
> So computationally you'd be adding a non-zero amount of load by performing a lightweight hash before storing and during look-ups, but it could be worth it for peace of mind.  

### Tokens Remain After Expiration
It appears that under normal usage, the database will eventually fill up with expired tokens belonging to a user which may eventually degrade performance and potentially cause a denial of service attack. 

It appears as though you're returning a 'not found' when a token is expired anyways, so it shouldn't break functionality to delete the entry when initially performing expiration logic.

### Singular Session Tokens Cannot Be Invalidated
It appears that there is no API to be able to delete a singular instance of a session token, only to invalidate all sessions. It would be beneficial to have that ability in case of a token leak / cookie insecurity situation, but also to provide an API to perform this action when a user initiates a "Sign Out" process in the consumer app.

Additionally having this API would more easily enable the creation of a session table allowing users to granularly revoke access to individual sessions.

## Things to be aware of
### Session Cookie Best Practices
Because you're leaving the configuration of Plug.Session values up to the developer leveraging WAC, they could forget to set secure cookie settings (Secure, HTTPOnly, other cookie security best practices) - just something to be aware of regarding the security model of using this library and maybe worth mentioning it in usage docs to encourage folks to use them.

I know that sort of falls more into the `Plug.Session` usage best practices, but it never hurts to take every opportunity to educate folks to set things up securely.

### Potential Timing Issue w/ Token Expiration Logic
Something about how/where you're performing the session token expiration time check feels off to me, but I've admittedly had a hard time putting my finger on it. I wonder if in a theoretical distributed system setup you could have a pod that is running on a different machine clock time and stores an `inserted_at` value that would be different than the machine performing the expiration logic.

It's a bit of a stretch I know and one you would still have with other token based systems, but I think most of my discomfort comes from not having the token itself be the singular source of truth for metadata used to invalidate it (a la JWTs with their baked in expiration time) and instead relying on the ancillary data in the database to inform the path forward - just cause in a vacuum, the session token value by itself doesn't know it's expired without having the `inserted_at` value.

No action needed here, just sharing my thoughts 🙂 

## Things I could use more context on
### CSRF Protections
I was a little confused on the CSRF protections in place here, I wouldn't mind a little bit more of an explanation there. I thought that since this relies on Plug.Storage that it would require developers setup Plug.CSRFProtection in their pipelines as well but I also see you doing some fanciness with regards to injecting anti-CSRF tokens in the javascript.ex generator that I may not be understanding fully

type1fool commented:
Thank you, @houllette, for the review and detailed response. I'm working through these one at a time. 🔥 

## Session Tokens Stored in Plaintext

Noted. For now, I'll rely on `Plug.Session`'s security measures and expect that developers altering the default behavior have a good understanding and reason to do so.

## Tokens Remain After Expiration

I've added a `UserTokenCleaner` GenServer, which runs `Identity.delete_expired_tokens/0` at the interval defined in `UserTokenCleaner`'s state as `:interval_minutes`. By default, the process executes the cleanup every 10 minutes, but this can be overridden in `application.ex` and updated while the application is running.

After running `wac.install` in a fresh Phoenix app, I can confirm `UserTokenCleaner` starts successfully and deletes expired tokens at the specified interval. The API functions for manually deleting and updating the interval also work correctly.

```
[debug] QUERY OK source="user_tokens" db=1.6ms decode=0.8ms queue=0.6ms idle=1725.4ms
DELETE FROM "user_tokens" AS u0 WHERE (u0."inserted_at" <= $1) [~N[2023-09-15 19:40:22]]
↳ Passkeys.UserTokenCleaner.handle_cast/2, at: lib/passkeys/user_token_cleaner.ex:104
[info] [expired_tokens_deleted: 0]
[debug] QUERY OK source="user_tokens" db=0.5ms idle=1498.6ms
DELETE FROM "user_tokens" AS u0 WHERE (u0."inserted_at" <= $1) [~N[2023-09-15 19:41:21]]
↳ Passkeys.UserTokenCleaner.handle_cast/2, at: lib/passkeys/user_token_cleaner.ex:104
[info] [expired_tokens_deleted: 0]
[debug] QUERY OK source="user_tokens" db=2.6ms idle=498.5ms
DELETE FROM "user_tokens" AS u0 WHERE (u0."inserted_at" <= $1) [~N[2023-09-15 19:42:21]]
↳ Passkeys.UserTokenCleaner.handle_cast/2, at: lib/passkeys/user_token_cleaner.ex:104
[info] [expired_tokens_deleted: 0]
```

## Singular Session Tokens Cannot Be Invalidated

I've added `Identity.delete_token/1` to provide an API for deleting single tokens. The session table is a good idea, and I may open an issue to implement that UI.

One thing I'm considering is that the approach so far with `WebauthnComponents` excludes any kind of "remember me" functionality. My thinking is that Passkeys are quite low-friction, and should become even easier as support becomes more ubiquitous. So, these tokens are valid for 24 hours by default, and developers can easily modify the code to adjust session length.

This _does_ mean that users must authenticate at least once every 24 hours, or whatever interval the developer sets. I'd like to revisit this so that the token is refreshed as the user navigates and interacts with the application, but I need to do some research to get it right.

## Session Cookie Best Practices

My expectation is that developers will be adding WAC to new or existing Phoenix applications, which already have `Plug.Session` configured securely as far as I know.

This is the browser pipeline in `router.ex` after running `wac.install`:

```elixir
  pipeline :browser do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_live_flash
    plug :put_root_layout, html: {PasskeysWeb.Layouts, :root}
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug :fetch_current_user # <- added by wac.install
  end
```

If I need to provide additional guidance, I'm happy to do so, but I don't know exactly what I would add. My usage of [`Plug.Conn.get_session/2`](https://hexdocs.pm/plug/1.14.0/Plug.Conn.html#get_session/2) and [`Plug.Conn.put_session/3`](https://hexdocs.pm/plug/1.14.0/Plug.Conn.html#put_session/3) may be naive, but my understanding is that the Phoenix-generated code in `MyAppWeb.Endpoint` provides a strong baseline of session security.

It does seem that `same_site: "Secure` ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie)) in `@session_options` would be preferable, but I wonder about the implications to Phoenix apps. This makes me wonder why `@session_options` doesn't start with `http_only: true` and `same_site: "Secure"`. 🤔 

This is an area a bit out of my expertise, so feel free to correct any misunderstanding or point to 

## Potential Timing Issue w/ Token Expiration Logic

Since Ecto & Phoenix [default to `:naive_datetime`](https://hexdocs.pm/ecto/Ecto.Schema.html#timestamps/1-options), this seems like a sensible default for `WebauthnComponents`. Developers deploying to a distributed environment would probably prefer UTC, but it seems acceptable to require them to update the generated files to make the switch. I could add an option for timezone, but that feels out of scope for this branch.

## CSRF Protections

The `_csrf_token` you see in `Wac.Gen.Javascript` is just a pattern that the generator uses to inject the JS hooks. That token code is generated by `phx.new` in ``lib/some_app/components/layouts/root.html.heex` and `assets/js/app.js` ([docs](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html#socket/3-connect-info)).
